Raise or Rails.error.report if ActiveSupport::Cache is given an invalid expiration time

trevorturk opened the pull request:
### Summary

Note this is a work in progress and should not be merged -- feedback needed, and the CI failures are expected (explained below...)

This is a followup from #45842 where @byroot suggested we might use the new `Rails.error.report` system (#43472) to report a `handled` error instead of just logging, and perhaps put raising an exception behind a new config flag, which could be the new default for 7.1. 

While working on this new PR, I tested using the `assert_called` helper and found that the `#merged_options` method appears to be called multiple times (in the Rails tests, and when running this Rails branch against a real app.) 

With this in mind, I propose we revert #45842, as it will currently log multiple times if an invalid expiration time is found. Because of this issue, I also propose that we alter our approach for this PR to only introduce the new config flag which will raise an exception (no logging or `Rails.error.report`ing) since this will avoid the multiple logging/reporting issue for now. 

Separately, this might highlight an area to explore for possible optimizations. I'm not sure if attempting to memoize `#merged_options` or something along those lines would make sense, but it may be worth considering. 

The relevant Rails test failure:

```
Failure:
MemoryStoreTest#test_invalid_expiration_time_reports_handled_error_when_raise_on_invalid_cache_expiration_time_is_false [rails/activesupport/test/cache/behaviors/cache_store_behavior.rb:588]:
[#<ArgumentError: Cache expiration time is invalid>, {:handled=>true}].
Expected report to be called 1 times, but was called 4 times.
Expected: 1
  Actual: 4
```

Example from me debugging using a Rails app of mine running against this PR:

```ruby
Rails.cache.fetch(cache_key, expires_at: 1.hour.ago) do
  # expensive thing...
end

#<ArgumentError: Cache expiration time is invalid>, {:expires_in=>-3600.000023}
#<ArgumentError: Cache expiration time is invalid>, {:compress=>false, :compress_threshold=>1024, :expires_in=>-3600.000023}
# expensive thing...
#<ArgumentError: Cache expiration time is invalid>, {:compress=>false, :compress_threshold=>1024, :expires_in=>-3600.000023}
#<ArgumentError: Cache expiration time is invalid>, {:compress=>false, :compress_threshold=>1024, :expires_in=>-3600.000023}
#<ArgumentError: Cache expiration time is invalid>, {:compress=>false, :compress_threshold=>1024, :expires_in=>-3600.000023}
```

A few other loose ends:

* I think we should add the invalid expiration time to the error message (any feedback on the message is welcome)
* Any feedback on the config name is welcome (also, I'm not 100% sure I implemented this correctly, but I followed the docs/guide and prior art around adding new framework defaults)
* If we decide to keep the logging and/or `error.report`ing, I have an open question if we should follow the pattern from #43712, where we use both `ActiveSupport.error_reporter` and also `logger.error`. 

I'll happily work up any changes requested, thanks for your consideration!

byroot commented:
Unless I'm mistaken, `merged_options` is always called like this:

```ruby
        options = merged_options(options)
```

So I think if you were to remove the `expires_in` from the returned options, the multiple call issue should be solved.

Let's try that and if it doesn't work we'll advise.

> I propose we revert https://github.com/rails/rails/pull/45842

Agreed. I'll revert it now, and I'll make sure this PR go through so that improvement is lost. Fell free to ping me as much as necessary to get this through.

trevorturk commented:
@byroot I've updated the PR according to your feedback, rebased against `master` to avoid merge conflicts with the revert of #45842, and squashed down to a single commit, including a new changelog entry etc. (Note that I left the existing PR description in place, but I'm happy to update that if desired.)

I added a `logger` call following the prior art in #43712. Note I don't believe this logger is set by default, so the log is likely to be missed. I see similar prior art in #42566 but it's a bit inconsistent. Any advice here would be great, but I'm happy to move along and assume this is configured as expected for those in the know. 

I wanted to pause before merging, though, to consider your proposed fix of removing the `expires_in` key. This works and does fix the issue as discussed, but it occurred to me that this might introduce an undesirable/unexpected behavior change. 

Consider an existing app that's been inadvertently sending an invalid `expires_at` time in the past. Before this change, the cache write would fail (or at least a subsequent cache read would come up empty.) After this change, the cache entry would not have an expiration time set, and would live on "forever". Please do let me know what you think. Perhaps we can come up with a more clever fix, or perhaps it may be worth removing the report/log conditional branch and just introducing the error behind the config flag.

byroot commented:
> Before this change, the cache write would fail (or at least a subsequent cache read would come up empty.) After this change, the cache entry would not have an expiration time set, and would live on "forever". Please do let me know what you think.

I think you are right :)

byroot commented on activesupport/lib/active_support/cache.rb:
```diff
@@ -718,6 +719,17 @@ def merged_options(call_options)
             call_options[:expires_in] = (expires_at - Time.now) if expires_at
 
+            if call_options[:expires_in]&.negative?
+              error = ArgumentError.new("Cache expiration time is invalid")
```
Best practice is to include the invalid value in the error message. It often makes it much easier to debug.

byroot replied:
Also you want to clearly state it's invalid because it's negative

byroot commented on activesupport/test/cache/behaviors/cache_store_behavior.rb:
```diff
@@ -566,6 +566,28 @@ def test_expires_in_and_expires_at
+    ActiveSupport::Cache::Store.logger = ActiveSupport::Logger.new(nil)
+    key = SecureRandom.uuid
+    error = ArgumentError.new("Cache expiration time is invalid")
+    assert_called(ActiveSupport.error_reporter, :report, [error, handled: true]) do
+      assert_called(ActiveSupport::Cache::Store.logger, :error, ["#{error}: #{error.message}"]) do
```
I'm not a fan of using `assert_called` for this. We should have a block based test helper that adds a new error error reporter that collects the errors and either return them to you or makes some assertions on them.

e.g. 
```ruby
assert_reports_error(error, handled: true) do
  # ...
end
```

byroot reviewed (commented):
Some minor things, but looks good to me overall.

trevorturk commented:
@byroot thanks again for your feedback, I updated the PR according to your feedback. I think we might want a few more tweaks, and I can rebase/squash etc. 

> I think you are right :)

I'm not sure, but I take your smiley face to mean we should leave the conditional branch with logging in place? Let me know if you wanted that removed, though!

I provided more detail in the error message. I'm using the `freeze_time` test helper so I can make assertions (new to me) but let me know if you'd prefer some other way. 

I started working on a block helper as you suggested, but I realized we have prior art from your error reporter work itself. I gave it a shot to move the test helper class into its own file so it could be shared with this test. I think it may be acceptable like this, but if you'd prefer a block method I was thinking of adding another new file like we have `MethodCallAssertions` called `ErrorReporterAssertions` or something. If you'd like this changed, please let me know what you have in mind, and if you'd want to refactor `ErrorReporterTest` to use the helper as well. 

Thanks!

byroot commented on activesupport/test/cache/behaviors/cache_store_behavior.rb:
```diff
@@ -566,6 +566,37 @@ def test_expires_in_and_expires_at
+  def test_invalid_expiration_time_reports_and_logs_when_raise_on_invalid_cache_expiration_time_is_false
+    ActiveSupport::Cache::Store.raise_on_invalid_cache_expiration_time = false
+
+    ActiveSupport.error_reporter = ActiveSupport::ErrorReporter.new
```
The reason I suggested a block based helper, is that here you are leaking a global value. 

This `ErrorReporter` instance will replace whatever `ActiveSupport.error_reporter` was set to, and survive past this test.

With a block based helper, it's much easier to record the previous value and properly set it back in an `ensure`.

byroot reviewed (approved):
Looks great to me aside from the test leak. Once it's fixed that's good for me.

Oh and please squash your commits once you are done.

byroot commented on activesupport/test/cache/behaviors/cache_store_behavior.rb:
```diff
@@ -566,6 +566,37 @@ def test_expires_in_and_expires_at
+    freeze_time
+    key = SecureRandom.uuid
+    error = assert_raises(ArgumentError) do
+      @cache.write(key, "bar", expires_at: 1.minute.ago)
```
I think you could just pass `-60` directly to simplify the test. Gets rid of the freeze_time etc.

byroot commented on activesupport/test/cache/behaviors/cache_store_behavior.rb:
```diff
@@ -566,6 +566,37 @@ def test_expires_in_and_expires_at
   end
 
+  def test_invalid_expiration_time_raises_an_error_when_raise_on_invalid_cache_expiration_time_is_true
+    ActiveSupport::Cache::Store.raise_on_invalid_cache_expiration_time = true
```
I just noticed you have another leak here. Global config like this must be set back at the end of the test, otherwise it can lead to test order dependent failures which make CI flaky.

byroot commented on activesupport/test/cache/behaviors/cache_store_behavior.rb:
```diff
@@ -566,6 +566,37 @@ def test_expires_in_and_expires_at
+  end
+
+  def test_invalid_expiration_time_reports_and_logs_when_raise_on_invalid_cache_expiration_time_is_false
+    ActiveSupport::Cache::Store.raise_on_invalid_cache_expiration_time = false
```
This needs to be restored too.

byroot commented:
> should leave the conditional branch with logging in place?

Yes, it was agreement.

trevorturk committed:
Add raise_on_invalid_cache_expiration_time config to ActiveSupport::Cache::Store

trevorturk commented:
@byroot I think I addressed your feedback, also squashed down the commits and rebased to avoid the merge conflict on the changelog. 

I just made two local helper methods since I'm not sure if this is more generalizable (yet) but I'm happy to work up more changes if you like. 

I slightly prefer the `freeze_time` (I think this can be a block instead) since it tests `expires_at` instead of `expires_in` but not a big deal.

Let me know if you want more changes, otherwise this may be ready if CI passes. 

Thanks!

byroot reviewed (approved):
All good for me now, thanks!

I'll merge on green.

trevorturk commented:
> I'll merge on green.

I don't think the CI failures are my fault, but lemme know if you'd like me to do anything else!

byroot commented:
Nah, you just hit a leaky test that I'm yet to debug. All good!
